mautic/mautic
# Install command and service

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | N
| Related user documentation PR URL | 
| Related developer documentation PR URL | mautic/developer-documentation#122
| Issues addressed (#s or URLs) | Resolves #7389
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Feature to allow command line install of Mautic. The objective is to migrate all the install process of `InstallController` to a service called `InstallService` and add an `InstallCommand` called by `php app/console mautic:install` that would leverage the new service.

Might also be useful to export each install steps into a different service / command in case one would only want to execute specific steps (something like `mautic:install:database`, `mautic:install:admin`, etc...).

This feature was funded by @Monogramm

#### Steps to test this PR:
Two things will need to be tested with this PR: 
1. the new install command(s)
2. the regular install wizard

Mautic install:
1. Load up this PR and user composer to install deps: #7395
```bash
git clone --branch feature/install-command git@github.com:Monogramm/mautic.git mautic_install
cd mautic_install
composer install
```
2. Create an initial `local.php` config file with the database parameters. You can also define the mailer properties and admin properties in `local.php` with the same syntax expected by the command line options (do `php bin/console mautic:install --help` for the list of options)
```php
<?php
// Example local.php to test install (to adapt of course)
$parameters = array(
	// Do not set db_driver and mailer_from_name as they are used to assume Mautic is installed
	'db_host' => 'localhost',
	'db_table_prefix' => null,
	'db_port' => 3306,
	'db_name' => 'mautic',
	'db_user' => 'mautic',
	'db_password' => 'mautic',
	'db_backup_tables' => true,
	'db_backup_prefix' => 'bak_',
	'admin_email' => 'admin@yopmail.com',
	'admin_password' => 'mautic',
	'mailer_transport' => null,
	'mailer_host' => null,
	'mailer_port' => null,
	'mailer_user' => null,
	'mailer_password' => null,
	'mailer_api_key' => null,
	'mailer_encryption' => null,
	'mailer_auth_mode' => null,
);

```
3. Execute `php bin/console mautic:install http://your-mautic-domain.com` (see [documentation](https://github.com/mautic/developer-documentation/pull/122/files#diff-860e3ce2541f17effc8aebd6b8cf606bR55))
```
Mautic Install
==============

Parsing options and arguments...
0 - Checking installation requirements...
Missing optional settings:
  - [0] It is recommended to secure your installation with an SSL certificate (https). Starting February 2020 Google Chrome will stop sending third-party cookies in cross-site requests unless the cookies are secure. Tracking will stop working for Mautic instances running on HTTP. Use HTTPS.
  - [1] The <strong>memory_limit</strong> setting in your PHP configuration is lower than the suggested minimum limit of %min_memory_limit%. Mautic can have performance issues with large datasets without sufficient memory.
Ready to Install!
1 - Creating database...
1.1 - Creating schema...
1.2 - Loading fixtures...
2 - Creating admin user...
3 - Email configuration and final steps...

================
Install complete
================

```
4. Check install worked properly and the install wizard is not triggered
5. Execute `php bin/console mautic:install` again and verify that nothing happens:
```
> php bin/console mautic:install http://your-mautic-domain.com

Mautic already installed
```

Mautic wizard:
1. Load up this PR: #7395
2. Go to your Mautic address and execute install wizard
3. Check install worked properly and the install wizard is not triggered anymore

#### List deprecations along with the new alternative:
None

#### List backwards compatibility breaks:
None (?)
